scene/gui/file_dialog.cpp

```diff
@@ -851,22 +876,57 @@ void FileDialog::update_file_list() {
 		}
 
 		if (match && (filename_filter_lower.is_empty() || filename.to_lower().contains(filename_filter_lower))) {
-			const Ref<Texture2D> icon = get_icon_func ? get_icon_func(base_dir.path_join(filename)) : theme_cache.file;
-			file_list->add_item(filename, icon);
-			file_list->set_item_icon_modulate(-1, theme_cache.file_icon_color);
-
-			if (mode == FILE_MODE_OPEN_DIR) {
-				file_list->set_item_disabled(-1, true);
-			}
-			Dictionary d;
-			d["name"] = filename;
-			d["dir"] = false;
-			d["bundle"] = false;
-			file_list->set_item_metadata(-1, d);
-
-			if (filename_edit->get_text() == filename || match_str == filename) {
-				file_list->select(file_list->get_item_count() - 1);
+			FileInfo fi;
+			fi.name = filename;
+			fi.match_string = match_str;
+
+			// Only assign sorting fields when needed.
+			if (file_sort == FileSortOption::TYPE || file_sort == FileSortOption::TYPE_REVERSE) {
+				fi.type_sort = filename.get_extension() + filename.get_basename();
+			} else if (file_sort == FileSortOption::MODIFIED_TIME || file_sort == FileSortOption::MODIFIED_TIME_REVERSE) {
+				fi.modified_time = FileAccess::get_modified_time(filename);
```

Comment:
This needs to be the path not the filename, or it won't work in folders.
Same with dir.

Reply:
It is a path.